### PR TITLE
Run kyuubi on kubernetes - make basic dockerfile and example

### PR DESCRIPTION
**Pull request by yaooqinn**
### _Why are the changes needed?_

```
# Usage:
# Start:
#   kubectl apply -f https://github.com/NetEase/kyuubi/blob/master/docker/example.yml
# Connect:
#   kubectl port-forward kyuubi-example --address localhost 10009:10009
# Or
#   kubectl exec -it kyuubi-example -- /bin/bash
#
#   ${SPARK_HOME}/bin/beeline -u 'jdbc:hive2://localhost:10009'
# Logging
#   kubectl logs -f kyuubi-example
```

### _How was this patch tested?_
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.readthedocs.io/en/latest/tools/testing.html#running-tests) locally before make a pull request

kubectl apply -f https://raw.githubusercontent.com/yaooqinn/kyuubi/docker/docker/example.yaml

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc @zwangsheng @pan3793 @ulysses-you @turboFei @zhaomin1423 @hddong thanks

**Comment by yaooqinn** [comment]
also cc @Yikun

File: docker/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    mkdir ${KYUUBI_PID_DIR} && \
+    mkdir ${KYUUBI_WORK_DIR_ROOT} && \
+    chmod ug+rw -R ${KYUUBI_HOME} && \
+    chmod a+rwx -R ${KYUUBI_WORK_DIR_ROOT} && \
+    rm -rf /var/cache/apt/*
+
+COPY bin ${KYUUBI_HOME}/bin
+COPY conf ${KYUUBI_HOME}/conf
+COPY jars ${KYUUBI_HOME}/jars
+COPY externals/engines/spark ${KYUUBI_HOME}/externals/engines/spark
```
Comment (Yikun) [inline docker/Dockerfile:55]: looks like `jars` and `externals/engines/spark` are not exists, does it require some pre-cmd? maybe we could add some doc in header doc.

File: docker/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+WORKDIR ${KYUUBI_HOME}
+
+CMD [ "/opt/kyuubi/bin/kyuubi", "run" ]
```
Comment (Yikun) [inline docker/Dockerfile:59]: Maybe:

```suggestion
CMD [ "sh", "-c", "$KYUUBI_HOME/bin/kyuubi", "run" ]
```

ref: https://github.com/moby/moby/issues/5509#issuecomment-45243772

File: docker/example.yaml [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# Usage:
+# Start:
+#   kubectl apply -f https://raw.githubusercontent.com/NetEase/kyuubi/blob/master/docker/example.yaml
```
Comment (Yikun) [inline docker/example.yaml:20]: Remove blob

```suggestion
#   kubectl apply -f https://raw.githubusercontent.com/NetEase/kyuubi/master/docker/example.yaml
```

File: docker/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+# limitations under the License.
+#
+
+# Usage:
+#   docker build -f docker/Dockerfile -t yaooqinn/kyuubi:tagname .
+#   Options:
+#     -f this docker file
+#     -t the target repo and tag name
+#     more options can be found with -h
+# TODO: REPALCE it with offical spark image iff Apache Spark community deploy
```
Comment (Yikun) [inline docker/Dockerfile:24]: typo: iff

File: docker/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+RUN set -ex && \
+    sed -i 's/http:\/\/deb.\(.*\)/https:\/\/deb.\1/g' /etc/apt/sources.list && \
+    apt-get update && \
+    apt install -y bash tini libc6 libpam-modules krb5-user libnss3 procps && \
+    useradd -u ${kyuubi_uid} -g root kyuubi && \
+    mkdir ${KYUUBI_HOME} && \
+    mkdir ${KYUUBI_LOG_DIR} && \
+    mkdir ${KYUUBI_PID_DIR} && \
+    mkdir ${KYUUBI_WORK_DIR_ROOT} && \
```
Comment (Yikun) [inline docker/Dockerfile:47]:
```suggestion
    mkdir -p ${KYUUBI_HOME} && \
    mkdir -p ${KYUUBI_LOG_DIR} && \
    mkdir -p ${KYUUBI_PID_DIR} && \
    mkdir -p ${KYUUBI_WORK_DIR_ROOT} && \
```

Add `-p` to make parent directories create as expected.

File: docker/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+COPY bin ${KYUUBI_HOME}/bin
+COPY conf ${KYUUBI_HOME}/conf
+COPY jars ${KYUUBI_HOME}/jars
+COPY externals/engines/spark ${KYUUBI_HOME}/externals/engines/spark
+
+WORKDIR ${KYUUBI_HOME}
+
+CMD [ "/opt/kyuubi/bin/kyuubi", "run" ]
+
+USER ${kyuubi_uid}
```
Comment (Yikun) [inline docker/Dockerfile:61]: that means we will run `/opt/kyuubi/bin/kyuubi run` use root, does it meet your expectations?

File: docker/Dockerfile (hunk 5 above)
Comment (Yikun) [reply docker/Dockerfile:47]: mkdir -p ${KYUUBI_HOME} ${KYUUBI_LOG_DIR} ${KYUUBI_PID_DIR} ${KYUUBI_WORK_DIR_ROOT}

File: docker/Dockerfile (hunk 4 above)
Comment (yaooqinn) [reply docker/Dockerfile:24]: iff means if and only if

File: docker/Dockerfile (hunk 6 above)
Comment (yaooqinn) [reply docker/Dockerfile:61]:
```
kubectl exec -it kyuubi-example -- whoami
kyuubi
```

Comment (pan3793) [reply docker/Dockerfile:59]: kubectl exec -it kyuubi-example -- ps aux

Comment (yaooqinn) [reply docker/Dockerfile:59]:
```
kubectl exec -it kyuubi-example -- ps aux
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
kyuubi       1  0.0  0.0   2288   696 ?        Ss   14:04   0:00 /usr/bin/tini -
kyuubi      15  0.0  0.0   5624  3296 ?        S    14:04   0:00 bash ./bin/kyuu
kyuubi      31  0.5  1.5 5395044 161284 ?      Sl   14:04   0:09 /usr/local/open
kyuubi      71  0.0  0.0   9396  3100 pts/0    Rs+  14:30   0:00 ps aux
```

File: docker/Dockerfile (hunk 4 above)
Comment (Yikun) [reply docker/Dockerfile:25]: learned, lol

File: docker/Dockerfile (hunk 6 above)
Comment (Yikun) [reply docker/Dockerfile:59]: OK, my bad understand on https://docs.docker.com/engine/reference/builder/#user. So, looks like CMD is regared as "for any RUN, CMD and ENTRYPOINT instructions that **follow** it in the Dockerfile"

**Review by ulysses-you — DISMISSED** [review]

**Pull request closed without merging**